app/src/main/java/org/apache/taverna/mobile/ui/playerlogin/PlayerLoginFragment.java

```diff
@@ -0,0 +1,263 @@
+    }
+
+
+    @OnClick(R.id.bLogin)
+    public void login(View v) {
+        if (ConnectionInfo.isConnectingToInternet(getContext())) {
+            if (!mEditTextEmail.getText().toString().trim().isEmpty() && !mEditTextPassword
+                    .getText().toString().trim().isEmpty()) {
+
+                mPlayerLoginPresenter.playerLogin(mEditTextEmail.getText().toString().trim(),
+                        mEditTextPassword.getText().toString().trim(), mCheckBoxRemember
+                                .isChecked());
+
+            } else {
+
+                showError("Please enter valid credential");
+            }
+        } else {
+
+            showError("NO Internet Connection");
+        }
+    }
+
+    @Override
+    public void showError(String string) {
+        final Snackbar snackbar = Snackbar.make(mEditTextPassword, string, Snackbar
+                .LENGTH_LONG);
+        snackbar.setAction("OK", new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                snackbar.dismiss();
+            }
+        });
+
+        snackbar.show();
+    }
+
+    @Override
+    public void showCredentialError() {
+        mTextInputEmail.setError(getString(R.string.err_login_email));
+        mTextInputPassword.setError(getString(R.string.err_login_password));
+        requestFocus(mEditTextPassword);
+    }
+
+    @Override
+    public void validCredential() {
+        mCallback.onSuccessfulLogin();
+    }
+
+
+    private void validateEmail() {
+
+        if (mEditTextEmail.getText().toString().trim().isEmpty()) {
+
+            mTextInputEmail.setError(getString(R.string.err_login_email));
+        } else {
+
+            mTextInputEmail.setError(null);
+        }
+
+
+    }
+
+    private void requestFocus(View view) {
+        if (view.requestFocus()) {
+            getActivity().getWindow().setSoftInputMode(WindowManager.LayoutParams
+                    .SOFT_INPUT_STATE_ALWAYS_VISIBLE);
+        }
+    }
+
+    private void validatePassword() {
+        if (mEditTextPassword.getText().toString().trim().isEmpty()) {
+
+            mTextInputPassword.setError(getString(R.string.err_login_password));
+        } else {
+
+            mTextInputPassword.setError(null);
+        }
+
+
+    }
+
+    @Override
+    public void onFocusChange(View v, boolean hasFocus) {
+        switch (v.getId()) {
+            case R.id.etEmail:
+                if (!v.hasFocus()) {
+                    validateEmail();
+                }
+                break;
+            case R.id.etPassword:
+                if (!v.hasFocus()) {
+                    validatePassword();
+                }
+                break;
+        }
+    }
+
+    @Override
+    public void onAttach(Context context) {
+        super.onAttach(context);
+
+        try {
+            mCallback = (OnSuccessful) getActivity();
+        } catch (ClassCastException e) {
+            throw new ClassCastException(getActivity().toString()
```

Comment:
@ianwdunlop This is for checking the fragment attached to Activity or not, If Activity is not implemented the interface then ClassCastException will be raised.

This is the architecture of fragment. 

For more details you can read this one: https://developer.android.com/training/basics/fragments/communicating.html and if you want to read in deep, please look : https://developer.android.com/guide/components/fragments.html